### DISCUSSION TITLE
Decorators as the recommendation

### DISCUSSION THREAD

**Discussion by bskimball** [sequelize/sequelize#16168] [Help & Questions] [answered]
I have been using Sequelize 7 alpha mostly because of IBMi odbc support. It has been working well and really like the syntax of extending the Model. 

I was quite surprised to log into the documentation today and see this "Sequelize also supports [a legacy approach](https://sequelize.org/docs/v7/other-topics/legacy-model-definitions/) that does not require using decorators, but this is discouraged."

IMO, the decorators are much harder to reason about. The field name is buried down at the bottom. I just don't see how this is more user friendly than using what is now referred to as a "legacy approach"

**Comment by ephys** [comment, marked as answer]
There are a number of advantages to using the new approach. Some thanks to decorators, others thanks to the approach where we now store metadata about the model but only init the models once they are provided to the Sequelize constructor.

The decorators make models much easier to type. We have had many issues with typing: We typed `.init` based on the properties declared on the class, because it was easy to make a typo. The problem with that is that some attributes are added automatically by Sequelize so you only need to declare it's type but you don't actually need to init it. It's conditional and the name varies so people using typescript had to put a dummy attribute in `init` to mute the warning.

Similarly for TypeScript users, the type déclaration and the attribute init were separated, making it easy to type incorrectly.

Naming associations is optional, but this lead to many subtle issues. In Sequelize 7, all associations are always named. By using decorators, we automatically use the name of the decorated property as the name of the association, encouraging naming the association like a model property, and making sure associations are all named.

We can better support plugins and third party libraries. For instance, we split off validation to a separate library called `@sequelize/validators.js`, because we want to open the door for other validation libraries like Joi, Zod, class-validators. This makes validation harder to set up, unless it's done via a decorator.

TypeScript again: attribute getters and setters were notoriously hard to type. With the new approach, we can just use native getters and setters, which solves many type issues.

Instance Hook methods and instance validation methods can just use `this` to access the model they're validating because it's just a method on the model. This is fairly intuitive and reduces the learning curve of the ORM.

They're statically analyzable which is something I'd like to exploit eventually to write an eslint plugin that catches common mistakes, as well as maybe eventually provide a tool to generate migrations by analysing models.

The current models are Active Records. This is great for many applications and we have no intention of changing this. However, some applications need to be able to use the same models in different schemas and databases. We have a few hacky workarounds to make that work but we want to support the Data Mapper pattern to better support these use cases. This is what we call Repository Mode. The way this would work is that you'd decorate a class that does not extend Model. Instead you'd access methods via a Repository by calling `sequelize.getRepository(Model)`. This has the advantage of being able to use the same model class with different sequelize instances, because the class only holds data.

I'm sure I'm forgetting a bunch of other reasons

---

The current decorator approach also first registers metadata, and your models are initialized when you give the models to initialize to the Sequelize constructor.

While this is possible to do without decorators, and I have considered supporting a decorator-less solution that used this approach, I have not due to the issues listed above, as well as because maintaining both documentations would be extremely time consuming.

This approach solves the following problems:

- It is now possible to have model inheritance. We merge the metadata with the metadata of parent models before initializing the model.
- We can now have abstract models that only exist for inheritance. These models are never registered.
- This is a huge one: Defining associations with `.init` and `.define` was a pain. You had to first define each model, before you could define associations. We have received _so many_ help requests from people struggling to understand how to set up their models. The new metadata approach registers all of the relevant metadata (of both the attributes and the associations) immediately, and Sequelize takes care of initializing models and associations in the right order with one uniform API.
